### ISSUE TITLE
Clarification on first frame input in `Emu3WorldModelDataset` vs. paper description

### ISSUE THREAD

**Issue by xiao10ma** [baaivision/UniVLA#35]
Hi, Yuqi! Sorry to bother you again!

In `train/datasets.py`, for class Emu3WorldModelDataset
```python
        # video VLA
        if self.video_format == "interleave":
            text_prompt = self.tokenizer.bos_token + prompt
            image_tokens = image_tokens[0::self.action_frames,...]
            if self.use_gripper and "gripper_image" in scene:
                gripper_tokens = gripper_tokens[0::self.action_frames,...]
            
            sample_text = self.tokenizer(text_prompt, padding=False, return_token_type_ids=False, return_tensors="pt")
            sample_input_ids = sample_text["input_ids"][0]
            sample_attention_mask = sample_text["attention_mask"][0]

            labels = torch.full((self.tokenizer.model_max_length,), fill_value=-100, dtype=torch.long)
            for i in range(len(image_tokens)):
                image_prompt = self.format_video_prompt(image_tokens[i:i+1])
                if self.use_gripper and "gripper_image" in scene:
                    gripper_prompt = self.format_video_prompt(gripper_tokens[i:i+1])
                    image_prompt += gripper_prompt
                sample_img = self.tokenizer(image_prompt, padding=False, return_token_type_ids=False, return_tensors="pt")
                image_input_ids = sample_img["input_ids"][0]
                image_attention_mask = sample_img["attention_mask"][0]
                
                sample_input_ids = torch.cat([sample_input_ids, image_input_ids], dim=-1)
                sample_attention_mask = torch.cat([sample_attention_mask, image_attention_mask], dim=-1)
                labels[len(sample_input_ids)-len(image_input_ids):len(sample_input_ids)] = image_input_ids
            
            if self.args.apply_loss_on_only_vision:
                labels = torch.where(torch.logical_and(labels >= self.bov, labels <= self.eov), labels, self.args.ignore_index)
```

It seems that for the first frame, it's input is only the text. However, for the paper, you said:
> Given the current observation $L^1_v$ and the instruction $L^1_t$ , the world model need to predict future visual content.

<img width="1259" height="364" alt="Image" src="https://github.com/user-attachments/assets/75cc71a2-f28e-459e-aae5-98e9023742ed" />

Can you clarify that? Than you in advance!

I think a better way is to exclude the first frame image's loss during training?

**Comment by Robertwyq** [comment]
@xiao10ma 
Since this is an NTP(next-token-prediction) training, the sequence of L actually includes the subtasks of L-1, which is essentially an inclusive relationship. Enabling all losses includes the scenario where the training condition is applied to the first frame.

**Comment by xiao10ma** [comment]
hi, thank you for your apply!
 I understand your explanation, but whether exclude the first frame is a better choice?

 As far as I’m concerned, it’s two tasks: one is to learn to generate from only text; another is to refer an image and generate the following images. Obviously, the second way is easier to learn!
 马梓培
***@***.***




---- Replied Message ----
From ***@***.***> Date 10/31/2025 23:22 To baaivision/UniVLA ***@***.***> Cc Zipei ***@***.***>,
Mention ***@***.***> Subject Re: [baaivision/UniVLA] Clarification on first frame input in `Emu3WorldModelDataset` vs. paper description (Issue #35)  
Robertwyq left a comment (baaivision/UniVLA#35)
@xiao10ma
 Since this is an NTP(next-token-prediction) training, the sequence of L actually includes the subtasks of L-1, which is essentially an inclusive relationship. Enabling all losses includes the scenario where the training condition is applied to the first frame.

**Comment by Robertwyq** [comment]
@xiao10ma 
Since this is only for pretraining, we directly used full supervision here. It might be worth trying, though I didn’t conduct relevant experiments at the time.

**Comment by xiao10ma** [comment]
Thank you so much for your valuable discussion!
